Linux: Linking as static library fails due to unresolved math-symbols (`log@@GLIC_*` and friends)

Steps to reproduce:
- Clone current `master` branch
- Go into cloned directory and execute `mkdir build && cd build && cmake -DBUILD_STATIC_LIBS=ON .. && make`

Error message:
```
/usr/bin/ld: ../src/libtrexio.a(trexio.c.o): undefined reference to symbol 'log@@GLIBC_2.29'
/usr/bin/ld: /lib/x86_64-linux-gnu/libm.so.6: error adding symbols: DSO missing from command line
```

It seems like an explicit linking with the maths library via `-lm` is required.

--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -13,6 +13,8 @@ project(Trexio
 set(CMAKE_C_STANDARD 99)
 set(CMAKE_C_STANDARD_REQUIRED ON)
 
+list(APPEND CMAKE_MODULE_PATH "${PROJECT_SOURCE_DIR}/cmake")
+
 # Optional configure for developer mode to generate source code from org-mode files.
 option(TREXIO_DEVEL "TREXIO developer mode (for code generation)." OFF)
 option(TREXIO_TESTS "Whether to perform tests for TREXIO" ON)

--- a/cmake/standard_math_lib.cmake
+++ b/cmake/standard_math_lib.cmake
@@ -26,12 +26,8 @@ return (int) (log(*argv[0]) + sin(*argv[1]) + sqrt(*argv[2]));
     return()
   endif()
 
-  find_library(STD_MATH_LIB "m")
+  set(STD_MATH_LIB "m")
 
-  if (NOT STD_MATH_LIB)
-    message(FATAL_ERROR "Linking to std math library required but the library could not be found")
-  endif()
-  
   set(CMAKE_REQUIRED_LIBRARIES "${STD_MATH_LIB}")
   check_c_source_compiles("${TEST_SOURCE}" "${VAR_BASENAME}_WITH_STDMATH")
 

--- a/cmake/standard_math_lib.cmake
+++ b/cmake/standard_math_lib.cmake
@@ -12,7 +12,12 @@ function(target_link_std_math TARGET VISIBILITY)
 #include <math.h>
 
 int main(int argc, char **argv) {
-return (int) (log(*argv[0]) + sin(*argv[1]) + sqrt(*argv[2]));
+/* Here, we use log( (double) *argv[0]) to convert the address of the string into
+a double. This will give random values: it will not convert the string into a
+double. But it will make the code valid for the compiler, and will avoid
+optimizations by the compiler to enforce the need of linking with the math
+library. */
+return (int) (log( (double) *argv[0]) + sin( (double) *argv[1]) + sqrt( (double) *argv[2]));
 }
 "
   )

--- a/cmake/standard_math_lib.cmake
+++ b/cmake/standard_math_lib.cmake
@@ -1,0 +1,43 @@
+# This function checks if the C standard library includes the symbols for
+# math operations and if not, will try to locate the auxiliary library (libm on Unix)
+# and link the given target to it.
+function(target_link_std_math TARGET VISIBILITY)
+  include(CheckCSourceCompiles)
+
+  # Clear variable to not be affected by external code
+  set(CMAKE_REQUIRED_LIBRARIES "")
+
+  set(TEST_SOURCE
+"
+#include <math.h>
+
+int main(int argc, char **argv) {
+return (int) (log(*argv[0]) + sin(*argv[1]) + sqrt(*argv[2]));
+}
+"
+  )
+
+  set(VAR_BASENAME "${PROJECT_NAME}_COMPILES_C")
+
+  check_c_source_compiles("${TEST_SOURCE}" "${VAR_BASENAME}")
+
+  if ("${VAR_BASENAME}")
+    # No explicit linking to math library required
+    return()
+  endif()
+
+  find_library(STD_MATH_LIB "m")
+
+  if (NOT STD_MATH_LIB)
+    message(FATAL_ERROR "Linking to std math library required but the library could not be found")
+  endif()
+  
+  set(CMAKE_REQUIRED_LIBRARIES "${STD_MATH_LIB}")
+  check_c_source_compiles("${TEST_SOURCE}" "${VAR_BASENAME}_WITH_STDMATH")
+
+  if (NOT "${${VAR_BASENAME}_WITH_STDMATH}")
+    message(FATAL_ERROR "Program doesn't compile even though linked against '${STD_MATH_LIB}'")
+  endif()
+  
+  target_link_libraries("${TARGET}" "${VISIBILITY}" "${STD_MATH_LIB}")
+endfunction()

--- a/src/CMakeLists.txt
+++ b/src/CMakeLists.txt
@@ -58,6 +58,9 @@ set_target_properties(trexio PROPERTIES
 	PUBLIC_HEADER ${TREXIO_PUBLIC_HEADERS}
 	)
 
+include(standard_math_lib)
+target_link_std_math(trexio PRIVATE)
+
 # ============= CONFIGURE HDF5 ==============
 
 # By defaulti, TREXIO is configured with the HDF5 support.

